packages/yew/src/virtual_dom/vnode.rs

```diff
@@ -104,9 +106,46 @@ where
     }
 }
 
-impl<T: ToString> From<T> for VNode {
-    fn from(value: T) -> Self {
-        VNode::VText(VText::new(value.to_string()))
+macro_rules! generate_into_vtext_impls {
+    ($($ty:ty),+ $(,)?) => {
+        $(
+        impl From<$ty> for VNode {
+            fn from(value: $ty) -> Self {
+                VNode::VText(VText::new(value))
+            }
+        }
+        )*
+    };
+}
+
+#[rustfmt::skip]
+generate_into_vtext_impls!(AttrValue, &'static str, String, Rc<str>, &String);
+
+macro_rules! generate_to_string_vtext_impls {
+    ($($ty:ty),+ $(,)?) => {
+        $(
+        impl From<$ty> for VNode {
+            fn from(value: $ty) -> Self {
+                VNode::VText(VText::new(::std::string::ToString::to_string(&value)))
+            }
+        }
+        )*
+    };
+}
+
+#[rustfmt::skip]
+generate_to_string_vtext_impls!(
+    u8, u16, u32, u64, u128,
+    i8, i16, i32, i64, i128,
+    f32, f64,
+    bool,
+    usize, isize,
```

Comment:
This adds something that  https://github.com/yewstack/yew/pull/1637 tried to get rid of (for props). We don't want to be implicitly allocating `String`s.

With this change, if someone were to pass in a number, it would automatically get converted to `String`, incurring a heap allocation. That should be explicit, just like it is props

Reply:
I'm not sure I'm following... before when you passed a primitive like that into the html!{} macro it would call ToString::to_string on it :thinking: am I missing something?

Reply:
Alright sorry I misread what you said!!

Ok so I agree with you and I'm all in favor of removing those! If I do so I will need to update a lot of tests (I can do it, np).

In the current implementation of VNode, ToString is used and string are allocated implicitly. But personally I do think it's best to be explicit. Especially when it involves allocating strings on the heap like that.

(Though this idea is conflicting with what @WorldSEnder point of view)

Reply:
> With this change, if someone were to pass in a number, it would automatically get converted to String, incurring a heap allocation. That should be explicit, just like it is props

(so it's already how it works actually. I did not introduce that. I added this for compatibility because the tests were failing)

Reply:
To clarify my viewpoint: The current behavior is to convert via `T: ToString` which has a blanket `impl<T: Display> ToString for T`. Arguably, the `String` will get converted to a `JsString` soon after. That allocation can not be avoided. I wonder if it was possible to use the `Display` impl to go directly to the `JsString` but that allocation (being part of FFI) is neither surprising nor avoidable. That being said, if a user would have to write `.to_string()` manually, that would lead to double the work, actually, since the resulting `String` would again need to be converted to a `JsString` as part of FFI.

As an aside, I think it's worthwhile to at least support `fmt::Arguments` (which is currently automatic through `Display -> ToString`).

Tl;dr: Custom logic that avoids the current `String` allocation welcome, but pushing it down to the user will only result in *more* allocations.

Reply:
Alright, I didn't think about this JSString...

Why don't we use JsString inside the VText struct then? I think they are garbage collected so cheap to clone and it's not like the user would want to read the string inside a VText, right?

Reply:
JsString is `utf-16`, which is expensive to construct from Rust strings which are `utf-8`.
Currently, a JsString is only needed if the content of a VText is needed.

Also, it's not possible to construct JsString for SSR if target is not `wasm32`.

Reply:
> Currently, a JsString is only needed if the content of a VText has changed.

Do you mean: most of the time, using String like we do currently, is actually more optimal?

I would have thought that the string always gets converted to JsString to be passed to JS, no matter what.

Reply:
> Do you mean: most of the time, using Rust's utf8 string like we do currently, is actually more optimal?
> I would have thought that the string always gets converted to JsString to be passed to JS, no matter what.

One of the primary advantage of using Virtual DOM is to reduce JavaScript-calls when possible.

See:

https://github.com/yewstack/yew/blob/master/packages/yew/src/dom_bundle/btext.rs#L82
https://docs.rs/web-sys/latest/web_sys/struct.Text.html#method.set_node_value

Reply:
> Yew currently tries to reduce JavaScripts call by checking the equity at the Rust side before it is passed to `web_sys`.

I see...

When I made implicit-clone I was wondering if I should check for reference equality instead of checking for actual equality. In JS for example we mostly use reference equality (same object in memory) rather than deep equality. A bit like clone vs deep-clone.

What would happen if we used the idx of JsValue to check if two JsString are equal? I guess this would be fast. I don't know if it's doable, it's just a wild idea.

Reply:
I would add that the conversion of `String` to `JsString` is, behind the scenes by wasm-bindgen, with [`TextDecoder`](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode) which I guess is the fastest way cross-browser to do the `utf-8` -> `utf-16` conversion. A (crude and not so telling) benchmark of different ways to concat slices of memory in javascript, see [this perf test].

Now, this doesn't compare preparing a single utf-8 string in wasm and then converting vs. converting each format part individually, and concatenating strings on javascript side. Note also there seems to be some [magic going on](https://accidentallyquadratic.tumblr.com/post/142387131042/nodejs-left-pad) that avoids quadratic performance degradation when building strings by sequentially appending to it in most browsers.

In the end, that single short-lived allocation for one `String` containing the `utf-8` encoded text to display might be faster than individually converting the parts the value is made out of.

[this perf test]: https://perf.link/#eyJpZCI6ImQ5ZGN4N3AwcWN5IiwidGl0bGUiOiJTdHJpbmcgY29uY2F0YW5hdGlvbiIsImJlZm9yZSI6ImNvbnN0IG51bVN0cmluZ3MgPSAxMDA7XG5jb25zdCBidWZmZXJzID0gWy4uLm5ldyBBcnJheShudW1TdHJpbmdzKV1cbiAgLm1hcCgoKSA9PiB7XG4gICAgcmV0dXJuIG5ldyBUZXh0RW5jb2RlcigpLmVuY29kZShNYXRoLnJhbmRvbSgpLnRvU3RyaW5nKCksICd1dGYtOCcpXG4gfSk7XG5jb25zdCBkZWNvZGVyID0gbmV3IFRleHREZWNvZGVyKCd1dGYtOCcsIHsgaWdub3JlQk9NOiB0cnVlIH0pO1xuXG5jb25zdCBjb25jYXRSZWR1Y2UgPSAoc3RycykgPT4gc3Rycy5yZWR1Y2UoXG4gIChhLCBiKSA9PiBhICsgZGVjb2Rlci5kZWNvZGUoYi5idWZmZXIpLCAnJyk7XG5cbmNvbnN0IGNvbmNhdExvb3AgPSAoc3RycykgPT4ge1xuICBsZXQgcmVzdWx0ID0gJydcbiAgZm9yIChsZXQgaSA9IDA7IGkgPCBzdHJzLmxlbmd0aDsgaSsrKSB7XG4gICAgcmVzdWx0ICs9IGRlY29kZXIuZGVjb2RlKHN0cnNbaV0uYnVmZmVyKTtcbiAgfVxuICByZXR1cm4gcmVzdWx0O1xufVxuXG5jb25zdCBtYXBKb2luRGVjb2RlID0gKHN0cnMpID0%2BIHtcbiAgcmV0dXJuIHN0cnMubWFwKHMgPT4gZGVjb2Rlci5kZWNvZGUocy5idWZmZXIpKS5qb2luKCcnKTtcbn1cblxuY29uc3QgbWFwQ29uY2F0RGVjb2RlID0gKHN0cnMpID0%2BIHtcbiAgcmV0dXJuICcnLmNvbmNhdCguLi5zdHJzLm1hcChzID0%2BIGRlY29kZXIuZGVjb2RlKHMuYnVmZmVyKSkpO1xufSIsInRlc3RzIjpbeyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImNvbmNhdFJlZHVjZShidWZmZXJzKTsiLCJydW5zIjpbMjAwMCw3NjAwLDEwMjAwLDIxMDAwLDE2ODAwLDEzMjAwLDE3MDAwLDE1NDAwLDExODAwLDEyNDAwLDE5NDAwLDEwODAwLDE0MDAwLDE1NDAwLDEwNDAwLDExNDAwLDU4MDAsMjAwMDAsOTgwMCwxMTIwMCwyMTgwMCwxMTIwMCwxMTgwMCwxODIwMCwxMDgwMCwxNzQwMCwxMzgwMCwxNDAwMCwxMzAwMCwxNDYwMCwyMDIwMCwxNjIwMCwxODYwMCwxODQwMCwxMzQwMCwxMjAwMCw4NDAwLDE5NjAwLDEzODAwLDEwNDAwLDEzODAwLDE2NDAwLDExNjAwLDkyMDAsMjAwMDAsMTAyMDAsMTU0MDAsMjAyMDAsMjAyMDAsMTM2MDAsMTIwMDAsMTEyMDAsMTU0MDAsODAwMCw1ODAwLDkwMDAsOTgwMCwxMTgwMCwxNDAwMCwxMTIwMCwxODAwMCw1ODAwLDEyMDAwLDEwNDAwLDE2MDAwLDY4MDAsMTY2MDAsMTM0MDAsMTYyMDAsODYwMCwxMzAwMCw2ODAwLDk2MDAsMTk2MDAsMTc4MDAsMTE4MDAsODAwMCwxOTYwMCwyMDIwMCwxMjYwMCwxNTYwMCwxNDQwMCw5ODAwLDE4MDAwLDE0MjAwLDkyMDAsMTU4MDAsMTU2MDAsMjE0MDAsMTEwMDAsNzYwMCwxMzYwMCw4NDAwLDEyNjAwLDUyMDAsMTE4MDAsMTk4MDAsOTIwMCwxNDgwMCw2MjAwXSwib3BzIjoxMzI5MH0seyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImNvbmNhdExvb3AoYnVmZmVycyk7IiwicnVucyI6WzM2MDAsMTM0MDAsMTY2MDAsMTk2MDAsMTQwMDAsMTUwMDAsMTc4MDAsMTU4MDAsMTE2MDAsMjU2MDAsMjI0MDAsMTE0MDAsMTQ2MDAsMTY4MDAsNzgwMCwxODgwMCwxNTQwMCwxOTIwMCwxNjQwMCwxMjIwMCwxNjIwMCwxMzYwMCw5MjAwLDIzNDAwLDE4NjAwLDE0ODAwLDE5NjAwLDE2NjAwLDE2ODAwLDE1MDAwLDE5ODAwLDIzNDAwLDE1NDAwLDE2MDAwLDIzMjAwLDI0NDAwLDExMDAwLDIzODAwLDE2MDAwLDExMjAwLDE4ODAwLDE0ODAwLDIwNDAwLDE2NjAwLDIyNDAwLDEwMDAwLDE5NDAwLDIzMDAwLDIyMjAwLDE3NDAwLDEyNDAwLDE1MjAwLDE3ODAwLDEzNjAwLDExMjAwLDE2NDAwLDEyODAwLDEzNjAwLDEzODAwLDIwNjAwLDIxMDAwLDEzMjAwLDEyODAwLDEzNjAwLDk2MDAsMjAwMDAsMTk0MDAsMTg2MDAsMTMwMDAsMTk0MDAsMTgyMDAsMTE4MDAsMTE4MDAsMTU0MDAsMzU0MDAsMTI0MDAsMTA2MDAsMTUyMDAsMTkyMDAsMjA0MDAsMTk2MDAsMTc4MDAsMTY0MDAsMTM0MDAsMTE2MDAsMTQ4MDAsMjAwMDAsMjE0MDAsMTkyMDAsMTcyMDAsMTM4MDAsMTM0MDAsMTQ0MDAsMTU2MDAsMTMwMDAsMTQyMDAsMTg4MDAsMTUyMDAsMTkwMDAsMTcwMDBdLCJvcHMiOjE2NDUyfSx7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoibWFwSm9pbkRlY29kZShidWZmZXJzKSIsInJ1bnMiOlsxODAwLDcwMDAsODgwMCwxNzQwMCwxMzIwMCwxODgwMCwxNTgwMCwxNDQwMCw2NjAwLDExNjAwLDE0MDAwLDE1NDAwLDEyMjAwLDE0MDAwLDk4MDAsMTAyMDAsMTY0MDAsMTMwMDAsMTM4MDAsMTM4MDAsMTA4MDAsODQwMCw4ODAwLDE0NDAwLDEzMDAwLDg2MDAsMTMyMDAsMTE2MDAsMTMyMDAsOTQwMCwxODIwMCwxMTAwMCwxODQwMCwxMjgwMCwxMTIwMCwxNzIwMCwxMTgwMCwxNjQwMCw4NDAwLDE0ODAwLDE0NDAwLDEzMDAwLDg2MDAsODQwMCwxMTQwMCwxMTYwMCwxNDIwMCwxMTQwMCwxMjQwMCwxMjYwMCwxMTYwMCw4ODAwLDEzMDAwLDEzODAwLDc4MDAsODQwMCw4NjAwLDg4MDAsMTU4MDAsOTQwMCwxNTgwMCwxMTgwMCwxMjAwMCw5NjAwLDk4MDAsNTQwMCwxMjAwMCw5ODAwLDg4MDAsODYwMCwxMzIwMCwxMDAwMCwxMTIwMCw4NDAwLDE2NjAwLDEzNDAwLDg0MDAsMTE2MDAsMTc0MDAsMTI0MDAsMTYyMDAsMTY4MDAsMTgyMDAsOTgwMCwxMzAwMCw5NjAwLDEwMDAwLDE5NjAwLDg4MDAsMTM0MDAsNzYwMCw4NjAwLDg2MDAsODQwMCw5MDAwLDE0MDAwLDg2MDAsODgwMCwxNjAwMCwxNDIwMF0sIm9wcyI6MTE4OTB9LHsibmFtZSI6IlRlc3QgQ2FzZSIsImNvZGUiOiJtYXBDb25jYXREZWNvZGUoYnVmZmVycykiLCJydW5zIjpbMjAwMCw3NDAwLDE1NDAwLDg0MDAsMTM2MDAsMTk2MDAsNjgwMCwxMDIwMCwxMDQwMCwxODYwMCwxNzYwMCwxMTQwMCwxNTYwMCwxNTYwMCw2NDAwLDcyMDAsMTM0MDAsMTg2MDAsMTI2MDAsMTI4MDAsMjA2MDAsOTYwMCw2NjAwLDEyMjAwLDYyMDAsMTU2MDAsMjIwMDAsOTAwMCwxMDQwMCw5NDAwLDE3MDAwLDgyMDAsMTE2MDAsMTM0MDAsMTM0MDAsODYwMCw5ODAwLDExMDAwLDExMjAwLDEzODAwLDg4MDAsMTEyMDAsMTQ2MDAsNzIwMCwxMTQwMCw5MjAwLDY4MDAsODgwMCwxNzIwMCwxMjQwMCw0ODAwLDEzNDAwLDEwMDAwLDQ4MDAsNDYwMCw3NDAwLDk4MDAsNjIwMCwxNjIwMCw5MjAwLDEzMjAwLDEwMDAwLDk2MDAsNjgwMCw0NjAwLDU4MDAsODYwMCw1ODAwLDEyMjAwLDUyMDAsMTE2MDAsODAwMCw5NjAwLDcwMDAsMzI4MDAsMTE0MDAsNjgwMCwxNjQwMCwxNTQwMCwxOTIwMCw4MDAwLDEwNDAwLDE5NjAwLDE0MjAwLDcyMDAsOTIwMCwxNDgwMCwxMzYwMCwxMzIwMCwxNDQwMCwxMDQwMCw2ODAwLDQ2MDAsNjAwMCw4NDAwLDQ0MDAsMTEyMDAsODAwMCwxMTAwMCw0NDAwXSwib3BzIjoxMDk3MH1dLCJ1cGRhdGVkIjoiMjAyMi0wNy0yOVQxMjoxNDoyMS4zNDlaIn0%3D

Reply:
Never mind, the user would not provide a JsString obviously :confounded:

Reply:
I have an(other) idea! :bulb: 

What if we make our own syntax for string formatting. Maybe something like this:

```
<div>format!(....)</div>
```
(notice the lack of `{}` around format! This is because it is not the actual format macro.

With that, instead of storing strings directly, we can store: the formatting string (static str) + the arguments (must implement ImplicitClone).

When the node is rendered, we compare all the arguments for equality.

This would avoid string heap allocation.

Reply:
@WorldSEnder @futursolo @hamza1311 I think I can make an implementation for this syntax:

```rs
fn compile_pass() {
    ::yew::html! { <div>format!("Hello {}!", "World")</div> };
}
```

It converts `format!(...)` to a `VFormat` type that holds all the arguments (`"Hello {}!"` and `"World"`) and does a `PartialEq` between those arguments and the ones from another VFormat.

```rs
// https://users.rust-lang.org/t/partialeq-eq-hash-any/70495
// https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c6be9c99223bb83fefc3094d42724f8b

#[derive(Clone)]
pub struct VFormat {
    arguments: Rc<Vec<Box<TypeTagged<dyn TaggedEq>>>>,
}

impl PartialEq for VFormat {
    fn eq(&self, other: &VFormat) -> bool {
        self.arguments == other.arguments
    }
}
```

However, it's rather complicated... if you don't mind I would prefer to do this in a separate PR and get this one merged. If this PR has no impact on performance at the moment, I think it's safe to merge. Let me know what you think.

Reply:
I think `format!` and `html!` are very similar and which both creates an owned instance based on formatting a certain representation with variables and can involve allocations if needed.

I am in favour of merging current implementation of `From<_> for VText`.

>     arguments: Rc<Vec<Box<TypeTagged<dyn TaggedEq>>>>,

This means that for each VFormat, we have an allocation overhead of:
1. Rc<_>.
2. Vec<_> for each argument.
3. Box<_> for each argument.

I feel it might be faster to use `std::format!` as it involves a single allocation overhead of creating a String.

In addition, types that implements `fmt::Display` are not required to implement `PartialEq` (nor does `PartialEq` sound for certain types like `f64`).

Reply:
Very good point... I think the VFormat thingy will reduce string formatting and conversion to JS but will increase the number of allocations. I'm not sure it's worth it.

Reply:
@futursolo I think I can reduce this to a single box for the entirety of the arguments. (But that won't solve the issue where PartiallEq is unsound (like for floats).